explorer/interpreter/type_checker.cpp

```diff
@@ -6122,7 +6122,13 @@ auto TypeChecker::DeclareDeclaration(Nonnull<Declaration*> d,
       CARBON_ASSIGN_OR_RETURN(
           Nonnull<const Value*> mixin,
           InterpExp(&mix_decl.mixin(), arena_, trace_stream_));
-      mix_decl.set_mixin_value(cast<MixinPseudoType>(mixin));
+      if (const auto* mixin_value = dyn_cast<MixinPseudoType>(mixin);
+          mixin_value) {
+        mix_decl.set_mixin_value(mixin_value);
+      } else {
+        return ProgramError(mix_decl.source_loc())
+               << "Not a valid mixin: `" << mix_decl.mixin() << "`";
+      }
```

Comment:
It seemed odd to me to use the implicit-bool but with a separate condition expression here... But maybe the whole thing would be a bit more idiomatic looking at other code in this file with an early return style, and an explicit null test?

```suggestion
      const auto* mixin_value = dyn_cast<MixinPseudoType>(mixin);
      if (mixin_value == nullptr) {
        return ProgramError(mix_decl.source_loc())
               << "Not a valid mixin: `" << mix_decl.mixin() << "`";
      }
      mix_decl.set_mixin_value(mixin_value);
```

Reply:
It's a mistake to have the mixin_value as it is, but I think if you read through the code you'll find we rarely do explicit null comparisons. e.g.:

https://github.com/carbon-language/carbon-lang/blob/trunk/explorer/interpreter/type_checker.cpp#L317
https://github.com/carbon-language/carbon-lang/blob/trunk/explorer/interpreter/type_checker.cpp#L414
https://github.com/carbon-language/carbon-lang/blob/trunk/explorer/interpreter/type_checker.cpp#L927-L941

And so on.

Lacking an explicit decision to change the entire codebase, I think it's better to be consistent with surrounding code.

Reply:
Not actually deliberate, but #2763 was because of me looking at dyn_cast usage for you; I noticed the buggy use that should've been just cast. It turns out the one nullptr comparison was for a variable that wasn't used, so isa is more appropriate.